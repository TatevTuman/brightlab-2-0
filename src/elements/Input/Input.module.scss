.input {
  display: flex;
  flex-direction: column;
  margin-bottom: 1rem;

  input {
    border-radius: 4px;
    border: 2px solid var(--primary-dark);
    padding: 0.5rem 1rem;
    transition: all 0.2s ease;

    &:active, &:focus {
      background-color: var(--primary-dark);
      color: var(--white);
      border-color: var(--primary-dark);
    }
  }
}
