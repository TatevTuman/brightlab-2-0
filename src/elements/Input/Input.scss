@import 'static/assets/styles/mixins';

.primary-input {
  @include disabled($background: false, $opacity: true, $pointerEvents: true);

  display: flex;
  flex-direction: column;
  width: 100%;

  &-inner {
    position: relative;
    display: flex;
    align-items: center;
    width: 100%;
    height: toRem(40);
    border-radius: 10px;
    transition: all 0.2s ease;
    background: var(--secondary-lst);
    border: 1px solid var(--secondary-lst);
    will-change: background, border-color;
    cursor: pointer;

    &:hover {
      background: var(--secondary-l);
      border: 1px solid var(--secondary-l);
    }

    &:active,
    &:focus-within {
      border: 1px solid var(--purple-o24);
      background: var(--white);
      outline: none;
    }

    input {
      flex: 1;
      height: 100%;
      border: none;
      overflow: scroll;
      font-size: toRem(15);
      color: var(--black);
      background: transparent;
      padding: 0.5rem 1rem;

      &[data-nocursor='true'] {
        text-shadow: 0 0 0 var(--black);
        color: transparent;
      }

      &::placeholder {
        color: var(--secondary);
      }
    }

    input:not(:placeholder-shown) + .primary-input-inner__suffix {
      .primary-input-inner__suffix-times {
        min-width: 16px;
        min-height: 12px;
        margin-right: 1rem;
        visibility: visible;
        opacity: 1;
      }
    }

    &__prefix {
      max-height: inherit;
      padding-top: 0.5rem;
      padding-bottom: 0.5rem;
      padding-left: 1rem;
      display: flex;
      align-items: center;
      justify-content: center;
      flex-shrink: 0;
      border-radius: inherit;
    }

    &__suffix {
      max-height: inherit;
      padding-top: 0.5rem;
      padding-bottom: 0.5rem;
      padding-right: 1rem;
      display: flex;
      align-items: center;
      justify-content: flex-start;
      overflow: hidden;
      border-radius: inherit;

      .loader {
        padding: 0;
        margin-right: 1rem;

        & > div {
          display: flex;
        }
      }

      &-times {
        width: 0;
        cursor: pointer;
        transition: all 0.3s ease;
        will-change: transform, opacity, visibility;
        visibility: hidden;
        opacity: 0;

        &:active {
          transform: scale(1.2);
        }
      }
    }
  }
}
