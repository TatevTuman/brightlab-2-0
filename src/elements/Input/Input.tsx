import React, { memo, forwardRef } from 'react'
import { handleEvent } from '@utils'
import TimesImage from '@images/times.svg'
import NotAllowedImage from '@images/not-allowed.svg'
import styles from './Input.module.scss'

export type InputSuffixProp = JSX.Element | string | number
export interface InputProps {
  value: string
  type?: string
  label?: string
  defaultValue?: string
  placeholder?: string
  prefix?: InputSuffixProp
  suffix?: InputSuffixProp

  direction?: 'horizontal' | 'vertical'
  autoComplete?: 'on' | 'off'
  disabled?: boolean
  error?: boolean
  clearable?: boolean
  required?: boolean
  focusable?: boolean
  role?: string
  min?: number

  onChange?: (value: string) => void
  onFocus?: (e: React.FocusEvent<HTMLInputElement>) => void
  onBlur?: (e: React.FocusEvent<HTMLInputElement>) => void
  onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => void
  onClear?: () => void
}

const Input = forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  const {
    value,
    type,
    label,
    defaultValue,
    placeholder,
    prefix,
    suffix,
    direction,
    autoComplete,
    disabled,
    error,
    clearable,
    required,
    focusable,
    role,
    min,
    onChange,
    onFocus,
    onBlur,
    onKeyDown,
    onClear
  } = props

  /* If disabled no focus */
  const tabIndex = focusable && !disabled ? 0 : -1

  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    handleEvent(onBlur, { value: e, disabled })
  }

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    handleEvent(onChange, { value: e.currentTarget.value, disabled })
  const handleFocus = (e: React.FocusEvent<HTMLInputElement>) => handleEvent(onFocus, { value: e, disabled })
  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => handleEvent(onKeyDown, { value: e, disabled })
  const handleClear = () => {
    if (!disabled) {
      onChange && onChange('')
    }
  }

  return (
    <div className={styles.input} data-disabled={disabled} data-direction={direction}>
      {label && (
        <label htmlFor={name} data-required={required}>
          {label}
        </label>
      )}
      <div className={styles.inputInner}>
        <input
          id={name}
          type={type}
          name={name}
          ref={ref}
          value={value}
          defaultValue={defaultValue}
          placeholder={placeholder}
          autoComplete={autoComplete}
          onChange={handleChange}
          onFocus={handleFocus}
          onKeyDown={handleKeyDown}
          onBlur={handleBlur}
          data-disabled={disabled}
          data-cursor={autoComplete !== 'off'}
          data-prefix={!!prefix}
          data-suffix={!!suffix}
          data-error={!!error}
          data-clearable={clearable}
          role={role}
          tabIndex={tabIndex}
          min={min}
        />
        <div className={styles.inputInnerPrefix}>{prefix}</div>
        <div className={styles.inputInnerSuffix}>
          {clearable && <TimesImage className={styles.inputInnerSuffixTimes} onClick={onClear || handleClear} />}
          {disabled && <NotAllowedImage className={styles.inputInnerSuffixNotAllowed} />}
          {suffix}
        </div>
      </div>
    </div>
  )
})

Input.defaultProps = {
  type: 'text',
  placeholder: 'Type a text',
  autoComplete: 'on',
  direction: 'vertical',
  focusable: true
}

Input.displayName = 'Input'

export default memo(Input)
