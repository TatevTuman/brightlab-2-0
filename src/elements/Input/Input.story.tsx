import React from 'react'
import { Meta, Story } from '@storybook/react/types-6-0'
import Input, { InputProps } from './Input'
import { storybook } from '@utils'

interface InputStoryTemplateProps extends InputProps {}

export default {
  title: 'Elements/Input',
  argTypes: {
    label: { control: 'text', name: 'Label' },
    placeholder: { control: 'text', name: 'Placeholder' },
    type: { control: storybook.controls.select(['text', 'number']), name: 'Type' },
    prefix: { control: 'text', name: 'Prefix' },
    suffix: { control: 'text', name: 'Suffix' },
    autoComplete: { control: storybook.controls.select(['on', 'off']), name: 'Autocomplete' },
    error: { control: 'boolean', name: 'Error' },
    clearable: { control: 'boolean', name: 'Clearable' },
    ...storybook.controls.interactive,
    onClear: { action: 'onClear' },
    name: storybook.args.disabled
  }
} as Meta

const InputStoryTemplate: Story<InputStoryTemplateProps> = args => {
  return <Input {...args} />
}

InputStoryTemplate.args = {
  ...storybook.actions.interactive,
  type: 'text',
  autoComplete: 'on',
  name: 'Input',
  clearable: true
}

export const InputStory = InputStoryTemplate.bind({})
InputStory.args = {
  ...InputStoryTemplate.args
}
