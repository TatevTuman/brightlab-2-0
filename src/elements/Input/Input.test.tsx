import React from 'react'
import { cleanup, render, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Input from './Input'

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('Input', () => {
  it('renders correctly', async () => {
    const { container } = render(<Input name={'input'} />)
    const awaitedContainer = await waitFor(() => container)

    expect(awaitedContainer).toMatchSnapshot()
  })

  it('renders no label correctly', () => {
    const { container } = render(<Input name={'input'} />)
    const label = container.querySelector('label')?.querySelector('label') || null

    expect(label).toBe(null)
  })

  it('renders label correctly', () => {
    const props = {
      name: 'input',
      label: 'input label'
    }

    const { getByText } = render(<Input {...props} />)

    expect(getByText(props.label)).toBeInTheDocument()
  })

  it('renders required label correctly', () => {
    const props = {
      name: 'input',
      label: 'input label',
      validation: { required: true }
    }

    const { getByText } = render(<Input {...props} />)
    const label = getByText(props.label)

    expect(label).toBeInTheDocument()
    expect(label).toHaveAttribute('data-required', 'true')
  })

  it('renders no validation error correctly', () => {
    const { container } = render(<Input name={'input'} />)
    const error = container.querySelector('.validation-error')

    expect(error).toBe(null)
  })

  it('renders validation error correctly', () => {
    const props = {
      name: 'input',
      errors: { input: { type: 'required', message: 'Input is required' } }
    }

    const { getByText } = render(<Input {...props} />)
    const error = getByText('Input is required')

    expect(error).toBeInTheDocument()
  })

  it('renders validation error correctly', () => {
    const props = {
      name: 'input',
      errors: { input: { type: 'required', message: 'Input is required' } }
    }

    const { getByText } = render(<Input {...props} />)
    const error = getByText('Input is required')

    expect(error).toBeInTheDocument()
  })

  it('change on label', () => {
    const props = {
      name: 'input',
      label: 'input label',
      onChange: jest.fn()
    }

    const { getByText, getByLabelText } = render(<Input {...props} />)
    const label = getByText(props.label)
    const input = getByLabelText(props.label) as HTMLInputElement

    userEvent.click(label!)
    userEvent.type(label!, 'test')

    expect(props.onChange).toHaveBeenCalledTimes(4)
    expect(input.value).toBe('test')
  })

  it('change', () => {
    const props = {
      name: 'input',
      label: 'input label',
      onChange: jest.fn()
    }

    const { getByLabelText } = render(<Input {...props} />)
    const input = getByLabelText(props.label) as HTMLInputElement

    userEvent.type(input!, 'test')

    expect(props.onChange).toHaveBeenCalledTimes(4)
    expect(input.value).toBe('test')
  })
})
