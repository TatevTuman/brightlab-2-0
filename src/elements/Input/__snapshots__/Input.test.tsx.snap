// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
<div>
  <div
    class="input"
    data-direction="vertical"
  >
    <div
      class="inputInner"
    >
      <input
        autocomplete="on"
        data-cursor="true"
        data-error="false"
        data-prefix="false"
        data-suffix="false"
        data-times="true"
        id="input"
        name="input"
        placeholder="Type a text"
        tabindex="0"
        type="text"
        value=""
      />
      <div
        class="inputInnerPrefix"
        data-prefix="false"
      />
      <div
        class="inputInnerSuffix"
        data-suffix="false"
      >
        <test-file-stub
          classname="inputInnerSuffixTimes"
        />
      </div>
    </div>
  </div>
</div>
`;
